src/api/helpers.jl

```diff
@@ -103,6 +103,49 @@ function h5a_iterate(@nospecialize(f), obj_id, idx_type, order, idx=0)
     return idxref[]
 end
 
+"""
+    h5a_iterate(Channel, loc_id, idx_type, order; kwargs...)
+    h5a_iterate(f, Channel, loc_id, idx_type, order; kwargs...)
+
+Create a `Channel{Tuple{hid_t,Ptr{Cchar},Ptr{H5A_info_t}}}` to iterate over the
+attributes. If a function `f` or a do block is provided the channel will be
+closed after the function or block returns.
+
+# Examples
+```julia-repl
+julia> HDF5.API.h5a_iterate(Channel, obj, HDF5.API.H5_INDEX_NAME, HDF5.API.H5_ITER_INC) do ch
+           for (loc, name, info) in ch
+               println(unsafe_string(name))
+           end
+       end
+```
+"""
+function h5a_iterate(::Type{Channel}, loc_id, idx_type, order, idx=0; kwargs...)
+    ch = Channel{Tuple{hid_t,Ptr{Cchar},Ptr{H5A_info_t}}}(0; kwargs...) do ch
+        h5a_iterate(loc_id, idx_type, order, idx) do loc, name, info
+            # release API lock
+            unlock(liblock)
+            try
+                isopen(ch) || return true
+                put!(ch, (loc, name, info))
```

Comment:
@vchuravy what happens when you `yield` from within a callback function?

Reply:
in particular, what happens to the outer `ccall` which invoked the callback?

Reply:
The outer `ccall` blocks but this is already running in another `Task` due to `@async`.

You'll also noticethat I release the API lock before calling `put!`. This allows the Julia code that uses `take!` to call other HDF5 procedures. I encountered a situation where the code breaks from the loop but does not close the channel. If I did not release the lock one would not be able to exit Julia since there is an exit hook to HDF5 which needs to acquire the lock.

Closing the channel should force a waiting `put!` command to error. The error should propagate to `HDF5.API.h5a_iterate_helper`, which will then terminate iteration.

Reply:
Yeah this should be fine. As long as HDF5 has no thread-local state itself.